src/safeposix/syscalls/fs_calls.rs

```diff
@@ -910,6 +909,421 @@ impl Cage {
+        }
+        return newfd;
+    }
+
+    //------------------------------------CLOSE SYSCALL------------------------------------
+
+    pub fn close_syscall(&self, fd: i32) -> i32 {
+        let fdtable = self.filedescriptortable.write().unwrap();
+        match fdtable.get(&fd) {
+            Some(_) => {return Self::_close_helper(self, fd, Some(&fdtable), true);},
+            None => {return syscall_error(Errno::EBADF, "close", "invalid file descriptor");},
+        }
+    }
+
+    pub fn _close_helper(&self, fd: i32, fdtable_lock: Option<&FdTable>, is_close_syscall: bool) -> i32 {
+        let writer;
+        let fdtable = if let Some(rl) = fdtable_lock {rl} else {
+            writer = self.filedescriptortable.write().unwrap(); 
+            &writer
+        };
+
+        if let Some(wrappedfd) = fdtable.get(&fd) {
+            let filedesc_enum = wrappedfd.read().unwrap();
+            let mut mutmetadata = FS_METADATA.write().unwrap();
+
+            //Decide how to proceed depending on the fd type.
+            //First we check in the file descriptor to handle sockets, streams, and pipes,
+            //and if it is a normal file descriptor we decrement the refcount to reflect
+            //one less reference to the file.
+            match &*filedesc_enum {
+                //if we are a socket, we dont change disk metadata
+                Stream(_) => {return 0;},
+                Socket(_) => {
+                    Self::_cleanup_socket(self, &fd, false);
+                 },
+                Pipe(pipe_filedesc_obj) => {
+                    let pipenumber = pipe_filedesc_obj.pipe;
+                    let read_references = Self::_lookup_refs_by_pipe_end(self, pipenumber, O_RDONLY);
+                    let write_references = Self::_lookup_refs_by_pipe_end(self, pipenumber, O_WRONLY);
+
+                    if write_references == 1 && pipe_filedesc_obj.flags == O_WRONLY {
+                        // let pipetable.pipenumber.eof = true;
+                    }
+
+                    if read_references + write_references == 1 {
+                        //del pipetable.pipenumber
+                    }
+                },
+                //TO DO: check IS_EPOLL_FD and if true, call epoll_object_deallocator (if necessary?)
+                File(normalfile_filedesc_obj) => {
+                    let inodenum = normalfile_filedesc_obj.inode;
+                    let inodeobj = mutmetadata.inodetable.get_mut(&inodenum).unwrap();
+                    let mut fobjtable = FILEOBJECTTABLE.write().unwrap();
+
+                    match inodeobj {
+                        Inode::File(ref mut normalfile_inode_obj) => {
+                            normalfile_inode_obj.refcount -= 1;
+
+                            //if there are still references to the file, then do nothing to the file
+                            if normalfile_inode_obj.refcount != 0 {
+                                return 0;
+                            }
```

Comment:
You still want to go through with removing the file descriptor from the fdtable even if the refcount is non-zero, the code will need to be refactored to allow that